Enhance image preprocessing pipeline

Adhamet opened the pull request:
- Added `default_mode` method to ensure input images are converted to the default BGR format.
- Improved `preprocess` function with new options for resizing, mode conversion (BGR, grayscale, black and white), noise reduction via median blur, and image normalization.
- Handled conversion for various image input types (NumPy arrays and PIL images) with flexible options for image processing.
- Ensured robust handling of image shape and color conversions.

Commit:
Enhance image preprocessing pipeline

Added `default_mode` method to ensure input images are converted to the default BGR format.
- Improved `preprocess` function with new options for resizing, mode conversion (BGR, grayscale, black and white), noise reduction via median blur, and image normalization.
- Handled conversion for various image input types (NumPy arrays and PIL images) with flexible options for image processing.
- Ensured robust handling of image shape and color conversions.

ahmedsaed committed:
fix type hints

ahmedsaed commented on smse/pipelines/image.py:
```diff
@@ -42,25 +42,80 @@ def load(self, input_path: Union[str, Path]) -> ImageT:
-    def preprocess(self, image: ImageT) -> ImageT:
-        """Preprocess image data"""
+    def default_mode(self, image: ImageT) -> ImageT:
+        """Converting image to default mode BGR"""
```
It might be better to refactor the function to handle different modes instead of a just BGR (it might not be the default for every model)

ahmedsaed commented on smse/pipelines/image.py:
```diff
@@ -42,25 +42,80 @@ def load(self, input_path: Union[str, Path]) -> ImageT:
+
+        return cast(ImageT, converted_image)
+
+    def preprocess(
+        self,
+        image: ImageT,
+        resizer: Union[bool, tuple[int, int]] = False,
+        mode_converter: int = 0,
+        median_blur: bool = False,
+        normalizer: bool = False,
+    ) -> ImageT:
```
Great Job! However, the function should get these parameters from the ImageConfig instead.

The function should take an image and process it based on the config provided at the initialization of the pipeline

ahmedsaed replied:
This function should also accept a batch of images or make a new function like `preprocess_batch` that utilizes `preprocess` to work on a batch of images

Commit:
Merge branch 'dev' of https://github.com/smse-org/SMSE into image-pipeline

Commit:
Merge branch 'dev' of https://github.com/smse-org/SMSE into image-pipeline

Commit:
Refactor image processing pipeline functions

- Updated 'color_mode' configuration logic to enhance flexibility in image conversions.
- Improved 'is_bgr' heuristic for determining BGR format with better intensity comparison logic.
- Enhanced 'default_mode' method to handle different image shapes and formats more robustly.
- Refactored 'mode_controller' to streamline conversions between color modes.
- Updated 'preprocess' function to include more user-defined flexibility in resizer, mode, and normalization.

Commit:
Refactor image processing pipeline functions

- Updated 'color_mode' configuration logic to enhance flexibility in image conversions.
- Improved 'is_bgr' heuristic for determining BGR format with better intensity comparison logic.
- Enhanced 'default_mode' method to handle different image shapes and formats more robustly.
- Refactored 'mode_controller' to streamline conversions between color modes.
- Updated 'preprocess' function to include more user-defined flexibility in resizer, mode, and normalization.

Commit:
Revert "Refactor image processing pipeline functions"

This reverts commit 0bf8425aed0e3c002ce83c485733b1763aff57ab.

Commit:
Add new image mode conversion capabilities

- Introduced 'mode_controller', 'default_mode', and 'is_bgr' methods for enhanced image mode conversion functionality.
- Updated 'ImageConfig' to include 'color_mode' attribute, enabling customizable image color modes, and removed the 'channels' attribute due to its limited utility.
- Modified the 'preprocess' method to support mode conversions via 'mode_controller' and 'default_mode'.
- Improved flexibility in handling image color conversions, including grayscale, RGB, BGR, and RGBA modes, while maintaining BGR heuristics for default handling.

These changes enhance image processing by enabling robust handling of various color modes and providing configurable normalization and resizing options.

ahmedsaed committed:
refactor pipelines

ahmedsaed committed:
refactor: move type hint definations to types.py

ahmedsaed committed:
lint

Commit:
Merge branch 'dev' of https://github.com/smse-org/SMSE into image-pipeline

Commit:
Applying torch in operations

ahmedsaed commented on smse/pipelines/image.py:
```diff
@@ -22,43 +37,66 @@ class ImagePipeline(BasePipeline):
+
+        return transforms.Compose(transform_list)
+
+    @staticmethod
+    def _ensure_rgb(image: Image.Image) -> Image.Image:
+        """Ensure image is in RGB mode."""
+        if image.mode != "RGB":
+            image = image.convert("RGB")
+        return image
```
This is hard coded for RGB, right? what if the user needs that images in BGR? 

You should also use native torchvision transforms instead of your custom one. I believe we can use [this transform](https://pytorch.org/vision/main/generated/torchvision.transforms.v2.functional.permute_channels.html?highlight=bgr) to handle the conversion between different modes

```
>>> rgb_image = torch.rand(3, 256, 256)
>>> bgr_image = F.permute_channels(rgb_image, permutation=[2, 1, 0])
```

We can define a set of permutations for each mode and apply the correct one based on the user's input.

There are also these transforms that might be handy
- RGB: https://pytorch.org/vision/main/generated/torchvision.transforms.v2.RGB.html?highlight=rgb#torchvision.transforms.v2.RGBhttps://pytorch.org/vision/main/generated/torchvision.transforms.v2.RGB.html?highlight=rgb#torchvision.transforms.v2.RGB
- GrayScale: https://pytorch.org/vision/main/generated/torchvision.transforms.v2.Grayscale.html?highlight=grayscale#torchvision.transforms.v2.Grayscale

ahmedsaed commented on smse/pipelines/image.py:
```diff
@@ -22,43 +37,66 @@ class ImagePipeline(BasePipeline):
+        Returns:
+            torch.Tensor: Preprocessed batch of images with shape [B, C, H, W].
+        """
+        if device is None:
+            device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
The base config already has a field called device which is set to `cpu` by default.

ahmedsaed commented on smse/pipelines/image.py:
```diff
@@ -22,43 +37,66 @@ class ImagePipeline(BasePipeline):
     def load(self, input_path: Union[str, Path]) -> ImageT:
         """Load image from file"""
         try:
             import cv2  # type: ignore[import-not-found]
-
             image: ImageT = cv2.imread(str(input_path))
-
+            if image is None:
+                raise ValueError(f"Failed to load image: {input_path}")
             return image
         except ImportError:
             self.logger.warning("OpenCV not found, trying PIL")
             from PIL import Image  # type: ignore[import-not-found]
-
             return np.array(Image.open(str(input_path)))
```
This function also has to be updated to support loading images from directory or a single image. It should also use native torchvision IO methods. check this out: https://pytorch.org/vision/main/io.html

Although, loading images is optional for now so you can ignore this comment.

Commit:
Flexible mode convertion and choice of device freedom

- added handling of various mode convertions for the user choice through config
- added the choice of the user for the device used whether cpu or gpu through cuda

Commit:
Re-factored load function to use PyTorch instead of CV

ahmedsaed reviewed (approved):
LGTM!
